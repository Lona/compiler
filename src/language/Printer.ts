import { doc, Doc } from 'prettier'
import { Definition, Value } from './Parser'
import { StateDefinition, Token, PrintPattern } from './Lexer'

const {
  builders: { concat },
} = doc

export function formatToken(
  lexerDefinition: StateDefinition[],
  token: Token
): Doc {
  const rules = lexerDefinition.flatMap(state => state.rules)

  const rule = rules.find(rule => rule.name === token.type)

  if (!rule) {
    throw new Error(`Couldn't find rule for token: ${token.type}`)
  }

  function formatPrintPattern(printPattern: PrintPattern): Doc {
    switch (printPattern.type) {
      case 'literal': {
        return printPattern.value
      }
      case 'indexReference': {
        return token.groups[printPattern.value]
      }
      case 'sequence': {
        return concat(printPattern.value.map(formatPrintPattern))
      }
    }
  }

  return rule.print ? formatPrintPattern(rule.print) : ''
}

export function formatTokens(
  lexerDefinition: StateDefinition[],
  tokens: Token[]
): Doc {
  return concat(tokens.map(formatToken.bind(null, lexerDefinition)))
}

export function print(
  document: Doc,
  options: doc.printer.Options = {
    printWidth: 80,
    tabWidth: 2,
    useTabs: false,
  }
) {
  return doc.printer.printDocToString(document, options).formatted
}
