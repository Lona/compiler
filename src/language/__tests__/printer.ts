import {
  Lexer,
  Rule,
  Action,
  rule,
  StateDefinition,
  LiteralPrintPattern,
  ReferencePrintPattern,
  PrintPattern,
  SequencePrintPattern,
} from '../Lexer'
import { formatTokens } from '../Printer'

function printLiteral(value: string): LiteralPrintPattern {
  return { type: 'literal', value }
}

function printReference(value: number): ReferencePrintPattern {
  return { type: 'reference', value }
}

function printSequence(value: PrintPattern[]): SequencePrintPattern {
  return { type: 'sequence', value }
}

it('prints simple tokens', () => {
  const definition: StateDefinition[] = [
    {
      name: 'main',
      rules: [
        rule('name', { pattern: '([a-z]+)', print: printReference(0) }),
        rule('equals', { pattern: '=', print: printLiteral('=') }),
        rule('not', { pattern: '!' }),
        rule('integer', {
          pattern: '(\\d)(\\d)(\\d)',
          print: printSequence([
            printReference(0),
            printReference(1),
            printReference(2),
          ]),
        }),
      ],
    },
  ]

  const lexer = new Lexer(definition)

  const tokens = lexer.tokenize('hello=!123')

  const doc = formatTokens(definition, tokens)

  expect(doc).toMatchSnapshot()
})
