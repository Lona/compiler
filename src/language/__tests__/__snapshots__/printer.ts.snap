// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints simple tokens 1`] = `
Object {
  "parts": Array [
    "hello",
    "=",
    "!",
    Object {
      "parts": Array [
        "1",
        "2",
        "3",
      ],
      "type": "concat",
    },
  ],
  "type": "concat",
}
`;
